chips/lowrisc/src/hmac.rs

```diff
@@ -0,0 +1,263 @@
+        let len = self.data_len.get();
+
+        if idx < len {
+            self.data.get().map(|data_ptr| {
+                let data_len = len - idx;
+
+                for i in 0..(data_len / 4) {
+                    if regs.status.is_set(STATUS::FIFO_FULL) {
+                        // Due to: https://github.com/lowRISC/opentitan/issues/1276
+                        //   we can't get back to processing the data as there is no
+                        //   FIFO not full interrupt.
+                        // Let's just keep going and put up with the back pressure.
+                        // break;
+                    }
+
+                    let data_idx = idx + (i * 4);
+
+                    unsafe {
+                        let mut d =
+                            (*(data_ptr.wrapping_offset(data_idx as isize + 0)) as u32) << 24;
+                        d |= (*(data_ptr.wrapping_offset(data_idx as isize + 1)) as u32) << 16;
+                        d |= (*(data_ptr.wrapping_offset(data_idx as isize + 2)) as u32) << 8;
+                        d |= (*(data_ptr.wrapping_offset(data_idx as isize + 3)) as u32) << 0;
+
+                        regs.msg_fifo.set(d);
+                    }
+
+                    self.data_index.set(data_idx + 4)
+                }
+
+                let idx = self.data_index.get();
+
+                for i in 0..(data_len % 4) {
+                    if regs.status.is_set(STATUS::FIFO_FULL) {
+                        // Due to: https://github.com/lowRISC/opentitan/issues/1276
+                        //   we can't get back to processing the data as there is no
+                        //   FIFO not full interrupt.
+                        // Let's just keep going and put up with the back pressure.
+                        // break;
+                    }
+
+                    let data_idx = idx + i;
+
+                    unsafe {
+                        let d = *(data_ptr.wrapping_offset(data_idx as isize)) as u32;
+
+                        regs.msg_fifo.set(d);
+                    }
+
+                    self.data_index.set(data_idx + 1)
+                }
+            });
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        let regs = self.registers;
+        let intrs = regs.intr_state.extract();
+
+        regs.intr_enable
+            .modify(INTR_ENABLE::HMAC_DONE::CLEAR + INTR_ENABLE::HMAC_ERR::CLEAR);
+
+        if intrs.is_set(INTR_STATE::HMAC_DONE) {
+            self.client.map(|client| {
+                client.hash_done(0);
+            });
+        } else if intrs.is_set(INTR_STATE::FIFO_FULL) {
+            // FIFO is full, we can't do anything
+        } else if intrs.is_set(INTR_STATE::HMAC_ERR) {
+            self.client.map(|client| {
+                client.hash_done(regs.err_code.get());
+            });
+        }
+    }
+}
+
+impl hil::hmac::HMAC<'a> for Hmac<'a> {
+    fn set_hash(&'a self, hash: HMACHash, key_size: usize) -> ReturnCode {
```

Comment:
nit: if I read the code correctly, `key_size` is not really needed: it's the size of a block for the given hashing algorithm. So it can be inferred, right?

I don't know OT implementations details but for a software HMAC implementation the key is not bounded: if it's bigger than a block for the selected hashing algorithm, the key is first hashed with this algorithm to make it fit in 1 block.